README.md
# Tailwind Inter Plugin

TailwindCSS Plugin to integrate with Inter Typeface from Rasmus Andersson [@rsms](https://twitter.com/rsms). It adds `.font-inter` class to apply Inter font family, adjusts letter spacing for each font size according to the [Dynamic Metrics](https://rsms.me/inter/dynmetrics/) and allows to toggle font feature settings. Optionally adds `@font-face` from [https://rsms.me/inter/inter.css](https://rsms.me/inter/inter.css).

The plugin is inspired with [tailwind-plugin-inter-font](https://github.com/suburbicode/tailwind-plugin-font-inter) plugin developed by Imam Susanto [@imsus](https://github.com/imsus).

## Installation

```sh
# with npm
npm install --save-dev tailwindcss-font-inter

# or with yarn
yarn add --dev tailwindcss-font-inter
```

## Usage

Add plugin to your `tailwind.config.js`:

```js
// tailwind.config.js
module.exports = {
    theme: {
        extend: {}
    },
    variants: {},
    plugins: [require('tailwindcss-font-inter')()]
};
```

Now you can put `.font-inter` class to apply the font (by default `@font-face` defenitions will be added to your CSS).

```html
<body class="font-inter text-base text-black bg-white antialiased font-feature-default">
    ...
</body>
```

You can set additional options for plugin:

```js
// tailwind.config.js
module.exports = {
  ...

  plugins: [
    require('tailwindcss-font-inter')({ // plugin's default settings
      a: -0.0223,
      b: 0.185,
      c: -0.1745,
      baseFontSize: 16,
      importFontFace: true,
    })
  ]
};
```

### Plugin Options

The plugin has several adjustable options.

-   `a`, `b`, `c` — constants used to calculate [Dynamic Metrics](https://rsms.me/inter/dynmetrics/).
-   `baseFontSize` — integer for base font size, default is `16`.
-   `importFontFace` — flag to inject Inter's `@font-face` to the output, default is `true`. If `false`, you should import Inter by yourself (e.g. from Google Fonts).

### Theme Configuration

The plugin uses [`fontSize` core plugin](https://tailwindcss.com/docs/configuration#core-plugins) to generate proper letter spacing for Inter Typeface. Additionally it uses `fontFeatureSettings` to config font feature settings.

```js
// tailwind.config.js
module.exports = {
    theme: {
        fontFeatureSettings: {
            numeric: ['tnum', 'salt', 'ss02']
        },
        fontSize: {
            xs: '0.75rem',
            sm: '0.875rem',
            base: '1rem',
            lg: '1.125rem',
            xl: '36px',
            '2xl': '1.5rem',
            '3xl': '1.875rem',
            '4xl': '2.25rem',
            '5xl': '3rem',
            '6xl': '4rem'
        }
    },
    plugins: [require('tailwindcss-font-inter')()]
};
```

### Font Family

The plugin will inject `@font-face` declaration from https://rsms.me/inter/inter.css (set `importFontFace` option to `false` to import font manually) and add `.font-inter` utility class.

## Font Sizes

Alongside with the default `text-lg` classes, the plugin will generate nested `text-lg` to set letter spacing according to Inter's **Dynamic Metrics**.

```css
.font-inter .text-xs,
.font-inter.text-xs {
    font-size: 0.75rem;
    letter-spacing: 0.0004908em;
}

.font-inter .text-sm,
.font-inter.text-sm {
    font-size: 0.875rem;
    letter-spacing: -0.0062235em;
}

.font-inter .text-base,
.font-inter.text-base {
    font-size: 1rem;
    letter-spacing: -0.0109598em;
}

.font-inter .text-lg,
.font-inter.text-lg {
    font-size: 1.125rem;
    letter-spacing: -0.0143007em;
}

.font-inter .text-2xl,
.font-inter.text-2xl {
    font-size: 1.5rem;
    letter-spacing: -0.0194923em;
}

.font-inter .text-3xl,
.font-inter.text-3xl {
    font-size: 1.875rem;
    letter-spacing: -0.0213145em;
}

.font-inter .text-4xl,
.font-inter.text-4xl {
    font-size: 2.25rem;
    letter-spacing: -0.0219541em;
}

.font-inter .text-5xl,
.font-inter.text-5xl {
    font-size: 3rem;
    letter-spacing: -0.0222574em;
}

.font-inter .text-6xl,
.font-inter.text-6xl {
    font-size: 4rem;
    letter-spacing: -0.0222974em;
}
```

### Font features

Also the plugin generates utility classes which allow you to specify named sets of font feature settings.

```js
// tailwind.config.js
module.exports = {
    theme: {
        interFontFeatures: {
            numeric: ['tnum', 'salt', 'ss02']
        }
    },
    plugins: [require('tailwindcss-font-inter')()]
};
```

This will generate the following classes:

```css
/* This is default classes */
.font-inter .font-feature-normal,
.font-inter.font-feature-normal {
    font-feature-settings: normal;
}

.font-inter .font-feature-default,
.font-inter.font-feature-default {
    font-feature-settings: 'calt' 1, 'kern' 1;
}

/* This is a custom class defined in config */
.font-inter .font-feature-numeric,
.font-inter.font-feature-numeric {
    font-feature-settings: 'tnum' 1, 'salt' 1, 'ss02' 1;
}
```

# License

[MIT](LICENSE.md)


CHANGELOG.md
### Version 2.0.3

- Fixed incorrect font size config type check (#13)

### Version 2.0.2

- Fixed Tailwind 2 fontSize default theme format incompatibility (#8)

### Version 2.0.1

- Refactoring styles generation
- Updated README.md
- Updated CHANGELOG.md
- tailwind 2 support

### Version 2.0.0

-   Inter font updated to v3.15
-   Fixed different size units caused `NaN` for letter spacing
-   Refactored styles generation
-   Removed line height usage as redundant
-   Removed options `baseLineHeight` and `disableUnusedFeatures`
-   Removed plugin specific font size classes to use inherited classes instead (`.text-inter-lg` → `.font-inter .text-lg`)
-   Limited font feature settings classes to apply only if Inter font family applied (`.font-feature-numeric` → `.font-inter .font-feature-numeric`)
-   Added `.font-feature-default` class with `calt` and `kern` font feature settings enabled
-   Option `importFontFace` is now `true` by default

### Version 1.0.8

-   Replaced Object.fromEntries with lodash methods to provide compatability older Node.js versions

### Version 1.0.7

-   Fixed line height and letter spacing values calculation for different font sizes (#2, #3)

### Version 1.0.6

-   Fixed npm package missed files

### Version 1.0.5

-   Bump config adjustments

### Version 1.0.4

-   Bump config adjustments

### Version 1.0.3

-   Bump config

### Version 1.0.2

-   Inter font updated to v3.11
-   Demo page updated
-   Small improvements

### Version 1.0.1

-   Fixed exception when no `interFontFeatures` config specified.

### Version 1.0.0

Currently supports fot feature settings and dynamic line height with the different font sizes.

-   Inter font updated to v3.10 (1.000)
-   Updated README.md


update.js
#!/usr/bin/env node

const { writeFile } = require('fs').promises;
const url = require('url');
const path = require('path');
const fetch = require('node-fetch');
const postcss = require('postcss');
const fontkit = require('fontkit');
const mkdirp = require('mkdirp').sync;
const rimraf = require('rimraf').sync;

const TEMP = path.resolve('.temp');

const toKebabCase = str =>
    str &&
    str
        .match(/[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g)
        .map(x => x.toLowerCase())
        .join('-');

const interUrl = 'https://rsms.me/inter/';
const interSource = interUrl + 'inter.css';
const interFamilies = new Set([]);
const interFiles = new Set([]);
const inter = {
    version: null,
    availableFeatures: new Set([]),
    base: {
        '@font-face': []
    },
    utilities: {
        '.font-inter': {
            'font-family': "'Inter', system-ui, sans-serif"
        },
        '@supports(font-variation-settings: normal)': {
            '.font-inter': {
                'font-family': "'Inter var', system-ui, sans-serif"
            }
        }
    }
};

const extractCss = root => {
    console.info('Parsing fetched CSS...');

    root.walkAtRules(rule => {
        let declarations = {};

        rule.walkDecls(decl => {
            let name = toKebabCase(decl.prop);
            let value = decl.value
                .split(',')
                .map(val => val.trim())
                .join(', ');

            if (name === 'font-family') {
                let fontName = value.replace(/^['"]|['"]$/g, '').trim();
                if (!fontName.match(/\salt$/)) {
                    interFamilies.add(fontName);
                }
            }

            if (name === 'src') {
                let vals = value.match(/url\(['"]([^'"]*)['"]\)/gi);

                if (vals) {
                    vals.forEach(val => {
                        val = val.replace(/^url\(['"]|['"]\)$/g, '');
                        value = value.replace(val, interUrl + val);
                        interFiles.add(interUrl + val);
                    });
                }
            }

            declarations[name] = value;
        });

        if (declarations['font-family'].match(/\salt\b/)) {
            console.log('Excluding declaration for font', declarations['font-family']);
            return;
        }

        inter.base[`@${rule.name}`].push(declarations);
    });

    console.log('Found font families:', [...interFamilies].join(', '));
};

const download = (fileUrl, file = null) => {
    if (!file) {
        const parsed = url.parse(fileUrl);
        const fileName = path.basename(parsed.pathname);
        file = path.join(TEMP, fileName);
    }

    require('child_process').execFileSync('curl', ['--silent', '-o', file, '-L', fileUrl], { encoding: 'utf8' });

    return file;
};

console.info('Fetching', interSource);

rimraf(TEMP);
mkdirp(TEMP);

fetch(interSource)
    .then(res => res.text())
    .then(css => postcss([extractCss]).process(css, { from: undefined }))
    .then(() => {
        console.info('Fetching font files...');

        for (fontFile of interFiles) {
            let file = download(fontFile);
            let font = fontkit.openSync(file);

            inter.version = font.version.replace(/Version\s/, '');
            font.availableFeatures.forEach(feature => inter.availableFeatures.add(feature));
        }

        inter.availableFeatures = [...inter.availableFeatures].sort();

        console.log('Font version:', inter.version);
        console.log('Found font features:', inter.availableFeatures.join(', '));
    })
    .then(async () => {
        let file = path.join(__dirname, 'inter.json');
        await writeFile(file, JSON.stringify(inter));
        console.info('Finished. Meta data stored in ./inter.json');
    })
    .catch(err => console.error(err));


src/utils.js
const parseUnit = require('parse-unit');

const PPI = 96;

const defaultUnitsRatio = {
    ch: 8,
    ex: 7.15625,
    em: 16,
    rem: 16,
    in: PPI,
    cm: PPI / 2.54,
    mm: PPI / 25.4,
    pt: PPI / 72,
    pc: PPI / 6,
    px: 1
};

const isNumeric = val => !isNaN(val) && !isNaN(parseFloat(val));

const toPx = (str, baseSize) => {
    if (!str && str !== 0) return null;
    if (defaultUnitsRatio[str]) return defaultUnitsRatio[str];

    // detect number of units
    const parts = parseUnit(str);

    if (isNumeric(parts[0])) {
        if (parts[1]) {
            if (parts[1] === '%') {
                return (parseFloat(parts[0]) / 100) * baseSize;
            }

            const px = toPx(parts[1], baseSize);
            return typeof px === 'number' ? parts[0] * px : null;
        } else {
            return parts[0];
        }
    }

    return null;
};

module.exports = {
    toPx
};


demo/scripts.js
import tailwindConfig from '../tailwind.config.js';
import resolveConfig from 'tailwindcss/resolveConfig';

const {
    theme: { interFontFeatures, fontSize }
} = resolveConfig(tailwindConfig);
const shortText = `
  <p class="font-black">Minute 360</p>
  <p class="font-semibold">Zenith zone</p>
  <p class="font-normal">1234567890!</p>
  <p class="font-thin">Grafik design</p>
`;
const longText = `
  <p class="whitespace-pre-wrap overflow-x-auto">ABCDEFGHIJKLMNOPQRSTUVWXYZ</p>
  <p class="whitespace-pre-wrap overflow-x-auto">abcdefghijklmnopqrstuvwxyz</p>
  <p class="whitespace-pre-wrap overflow-x-auto">0 1 2 3 4 5 6 7 8 9  7*4  7×4  3/4  7÷8  3°  ℃  ℉</p>
  <p class="whitespace-pre-wrap overflow-x-auto">64%  90px  45 kg   $64 $7   €64  £7  1 440 ₽</p>
  <p class="whitespace-pre-wrap overflow-x-auto">Å Ä Ö Ë Ü Ï Ÿ å ä â ö ë ü ï î ÿ Ø ø • ∞ ~</p>
  <p class="whitespace-pre-wrap overflow-x-auto">. ‥ … → ← ↑ ↓</p>
`;

console.log(interFontFeatures, fontSize);

const query = selector => document.querySelector(selector);
const render = str => {
    const el = document.createElement('div');
    el.innerHTML = str;
    return el.firstElementChild;
};

function printFontSizes(sizes = {}) {
    let container = query('#fontSizes');

    if (container) {
        Object.entries(sizes)
            .reverse()
            .forEach(([name]) => {
                const el = render(
                    `<div>
                      <p class="text-xs text-gray-500 py-2">
                        .text-inter-${name}
                      </p>
                      <div class="mb-10 text-${name}">
                        ${shortText}
                      </div>
                    </div>`
                );
                container.appendChild(el);
            });
    }
}

function printFontFeatures(features = {}) {
    let container = query('#fontFeatures');

    if (container) {
        Object.entries(features).forEach(([name, value]) => {
            const el = render(
                `<div>
                  <p class="text-xs text-gray-500 py-2">
                    .font-feature-${name} (${value.join(', ')})
                  </p>
                  <div class="mb-10 text-2xl font-feature-${name}">
                    ${longText}
                  </div>
                </div>`
            );
            container.appendChild(el);
        });
    }
}

printFontSizes(fontSize);
printFontFeatures(interFontFeatures);


src/index.js
const { toPx } = require('./utils');
const Inter = require('../inter.json');

const isBoolean = val => typeof val === 'boolean';
const isString = val => typeof val === 'string';
const isArrayLike = obj => obj != null && !isString(obj) && typeof obj[Symbol.iterator] === 'function';
const isPlainObject = val => !!val && typeof val === 'object' && val.constructor === Object;
const entries = obj => Object.keys(obj).map(key => [key, obj[key]]);
const fromEntries = iterable =>
    [...iterable].reduce((obj, [key, val]) => {
        obj[key] = val;
        return obj;
    }, {});
const mapObject = (obj, cb) => fromEntries((Array.isArray(obj) ? obj : entries(obj)).map(val => cb(...val)));
const filterObject = (obj, cb) => fromEntries((Array.isArray(obj) ? obj : entries(obj)).filter(val => cb(...val)));
const defaults = (obj, ...defs) => Object.assign({}, obj, ...defs.reverse(), obj);
const round = (num, prec = 3) => parseFloat(num.toFixed(prec));
const unquote = str => str.replace(/^['"]|['"]$/g, '').trim();
const tracking = (fontSize, a, b, c) => a + b * Math.pow(Math.E, c * fontSize);

const normalizeEntry = (key, value) => {
    value = isBoolean(value) ? '' + 1 * value : '' + value;
    value = value !== '1' && value !== 'undefined' ? value : '1';

    return [unquote(key), value];
};

const generateFeatures = (features, available) => {
    if (isPlainObject(features)) {
        features = mapObject(features, (key, value = '1') => normalizeEntry(key, value));
    } else {
        if (isString(features)) {
            features = fromEntries(features.split(',').map(f => f.trim().split(' ')));
        }

        features = fromEntries(
            features.map(feature => {
                let key, value;

                if (isString(feature)) {
                    [key, value = '1'] = feature.replace(/\s\s+/g, ' ').split(' ', 2);
                } else if (isArrayLike(feature)) {
                    [key, value = '1'] = feature;
                } else if (isPlainObject(feature)) {
                    [key, value = '1'] = entries(feature)[0];
                }

                return normalizeEntry(key, value);
            })
        );
    }

    features = filterObject(features, key => available.includes(key));

    return entries(features)
        .map(([key, value]) => `"${key}" ${value}`)
        .filter(val => !!val)
        .sort()
        .join(', ')
        .trim();
};

module.exports = function (options = {}) {
    return ({ addBase, addUtilities, variants, e, theme }) => {
        const { availableFeatures, utilities, base } = Inter;

        const defaultConfig = defaults(options, {
            a: -0.0223,
            b: 0.185,
            c: -0.1745,
            baseFontSize: 16,
            importFontFace: true
        });

        const defaultFontFeaturesTheme = { default: ['calt', 'kern'] };
        const defaultFontSizeVariants = ['responsive'];

        const fontSizeTheme = theme('fontSize', []);
        const fontSizeVariants = variants('fontSize', defaultFontSizeVariants);
        const fontFeaturesTheme = theme('interFontFeatures', defaultFontFeaturesTheme);

        const fontFeatures = defaults(fontFeaturesTheme, defaultFontFeaturesTheme);
        const baseStyles = {
            ...(defaultConfig.importFontFace ? base : {})
        };

        const fontSizeStyles = (fontSize, { a, b, c }) => {
            const size = isArrayLike(fontSize) ? fontSize[0] : fontSize;
            const sizeInPx = toPx(size, defaultConfig.baseFontSize);
            const trackingSize = tracking(sizeInPx, a, b, c);

            return {
                fontSize: size,
                letterSpacing: `${round(trackingSize, 7)}em`
            };
        };

        const fontFeatureStyles = value => {
            return value.length ? { fontFeatureSettings: Array.isArray(value) ? value.join(', ') : value } : null;
        };

        const fontFeatureUtilities = {
            ...{
                '.font-inter .font-feature-normal, .font-inter.font-feature-normal': {
                    ...fontFeatureStyles('normal')
                }
            },
            ...mapObject(fontFeatures, (modifier, value) => {
                const features = generateFeatures(value, availableFeatures);

                return [
                    `.font-inter .${e(`font-feature-${modifier}`)},.font-inter.${e(`font-feature-${modifier}`)}`,
                    {
                        ...fontFeatureStyles(features)
                    }
                ];
            })
        };

        const fontSizeUtilities = mapObject(fontSizeTheme, (modifier, value) => {
            const { a, b, c } = defaultConfig;

            return [
                `.font-inter .${e(`text-${modifier}`)}, .font-inter.${e(`text-${modifier}`)}`,
                {
                    ...fontSizeStyles(value, { a, b, c })
                }
            ];
        });

        // Add @font-face if importFontFace: true
        // see https://rsms.me/inter/inter.css
        addBase(baseStyles);

        // Add .font-inter
        addUtilities(utilities);

        // Add .font-feature-{modifier} utility classes
        addUtilities(fontFeatureUtilities);

        // Add .font-inter.text-{size} utility classes
        addUtilities(fontSizeUtilities, fontSizeVariants);
    };
};
